modules/runner/src/integrationTest/java/org/apache/ignite/internal/table/ItInternalTableTest.java

```diff
@@ -384,78 +359,190 @@ public void testRoScanAllExplicitPopulatingData() throws InterruptedException {
+
+                assertEquals(i % 100L, rowTuple.<Long>value("key"));
+                assertEquals(i, rowTuple.<Integer>value("valInt"));
+                assertEquals("some string row" + i, rowTuple.<Integer>value("valStr"));
             }
+        }
+    }
+
+    private ArrayList<BinaryRowEx> populateEvenKeysAndPrepareEntriesToLookup(boolean keyOnly) {
+        KeyValueView<Tuple, Tuple> keyValueView = table.keyValueView();
+
+        var keyRows = new ArrayList<BinaryRowEx>();
 
-            @Override
-            public void onComplete() {
-                subscriberAllDataAwaitLatch2.countDown();
+        for (int i = 0; i < 15; i++) {
+            keyRows.add(keyOnly ? createKeyRow(i) : createKeyValueRow(i, i, "some string row" + i));
+
+            if (i % 2 == 0) {
+                putValue(keyValueView, i);
             }
-        });
+        }
+
+        Collections.shuffle(keyRows);
 
-        subscriberAllDataAwaitLatch2.await();
+        return keyRows;
+    }
+
+    private void roScanAll(boolean implicit) throws InterruptedException {
+        IgniteImpl node = node();
+
+        KeyValueView<Tuple, Tuple> keyValueView = table.keyValueView();
+
+        var retrievedItems = scanAllPartitions(node);
+
+        assertEquals(0, retrievedItems.size());
+
+        populateData(node, keyValueView, implicit);
+
+        retrievedItems = scanAllPartitions(node);
 
         assertEquals(15, retrievedItems.size());
     }
 
+
+    /**
+     * Scans all table entries.
+     *
+     * @param node Ignite instance.
+     * @return Collection with all rows.
+     * @throws InterruptedException If fail.
+     */
+    private static List<BinaryRow> scanAllPartitions(IgniteImpl node) throws InterruptedException {
+        InternalTable internalTable = ((TableImpl) node.tables().table(TABLE_NAME)).internalTable();
+
+        List<BinaryRow> retrievedItems = new CopyOnWriteArrayList<>();
+
+        int parts = internalTable.partitions();
+
+        var subscriberAllDataAwaitLatch = new CountDownLatch(parts);
+
+        for (int i = 0; i < parts; i++) {
+            Publisher<BinaryRow> res = internalTable.scan(i, node.clock().now(), node.node());
+
+            res.subscribe(new Subscriber<>() {
+                @Override
+                public void onSubscribe(Subscription subscription) {
+                    subscription.request(10000);
+                }
+
+                @Override
+                public void onNext(BinaryRow item) {
+                    retrievedItems.add(item);
+                }
+
+                @Override
+                public void onError(Throwable throwable) {
+                    fail("onError call is not expected.");
+                }
+
+                @Override
+                public void onComplete() {
+                    subscriberAllDataAwaitLatch.countDown();
+                }
+            });
+        }
+
+        subscriberAllDataAwaitLatch.await();
```

Comment:
Waiting without a timeout is dangerous.

Reply:
The logic is just a copy of the previous one.

Reply:
Please fix this.